### PR TITLE
Add distributed commands related metrics to job service

### PR DESCRIPTION
**Pull request by luzhang6**
### What changes are proposed in this pull request?
Add new distributed command metrics, operation counts by status (success, fail, cancel), file count, file size metrics.

1. Tests run in local mode.
2. Add unit test for cancel operations.

Ran distributedLoad command and could see the counters change their values accordingly.
` ./bin/alluxio fs distributedLoad /data/1023
Allow up to 3000 active jobs
/data/1023 loading
Successfully loaded path /data/1023 after 1 attempts
Completed count is 1,Failed count is 0.`
    "JobMaster.Master.JobDistributedLoadCancel" : {
      "count" : 0
    },
    "JobMaster.Master.JobDistributedLoadFail" : {
      "count" : 0
    },
    "JobMaster.Master.JobDistributedLoadFileCount" : {
      "count" : 1
    },
    "JobMaster.Master.JobDistributedLoadFileSizes" : {
      "count" : 12
    },
    "JobMaster.Master.JobDistributedLoadSuccess" : {
      "count" : 1
    },
  "meters" : {
    "JobMaster.Master.JobDistributedLoadRate" : {
      "count" : 12,
      "m15_rate" : 0.01329636478824785,
      "m1_rate" : 0.1918934048896241,
      "m5_rate" : 0.039668510828118016,
      "mean_rate" : 0.0051907953589612026,
      "units" : "events/second"
    }
  },

Ran distributedCp command, and see the counters change below:
`./bin/alluxio fs  distributedCp /data/1023  /data/1023copy
Allow up to 3000 active jobs
Copying /data/1023 to /data/1023copy
Successfully copied /data/1023 to /data/1023copy after 1 attempts`
"JobMaster.Master.MigrateJobCancel" : {
      "count" : 0
    },
    "JobMaster.Master.MigrateJobFail" : {
      "count" : 0
    },
    "JobMaster.Master.MigrateJobFileCount" : {
      "count" : 1
    },
    "JobMaster.Master.MigrateJobFileSize" : {
      "count" : 12
    },
    "JobMaster.Master.MigrateJobSuccess" : {
      "count" : 1
    },

For Persist metrics after running runTests and manually loading or copying files (total 14 files):
"JobMaster.Master.AsyncPersistCancel" : {
      "count" : 0
    },
    "JobMaster.Master.AsyncPersistFail" : {
      "count" : 0
    },
    "JobMaster.Master.AsyncPersistFileCount" : {
      "count" : 14
    },
    "JobMaster.Master.AsyncPersistFileSize" : {
      "count" : 1059
    },
    "JobMaster.Master.AsyncPersistSuccess" : {
      "count" : 14
    },

Please outline the changes and how this PR fixes the issue.

### Why are the changes needed?

Please clarify why the changes are needed. For instance,
  1. If you propose a new API, clarify the use case for a new API.
  2. If you fix a bug, describe the bug.

### Does this PR introduce any user facing changes?

Please list the user-facing changes introduced by your change, including
  1. change in user-facing APIs
  2. addition or removal of property keys
  3. webui

### CODE REVIEW

File: job/server/src/main/java/alluxio/master/job/metrics/DistributedCmdMetrics.java
Comment (jja725) [inline job/server/src/main/java/alluxio/master/job/metrics/DistributedCmdMetrics.java:174]: why not use instanceof? And I think in this case we don't really need to change the source parameter for MigrateConfig

File: job/server/src/main/java/alluxio/master/job/metrics/DistributedCmdMetrics.java
Comment (jja725) [inline job/server/src/main/java/alluxio/master/job/metrics/DistributedCmdMetrics.java:228]: just use instanceof

File: job/server/src/main/java/alluxio/master/job/plan/PlanCoordinator.java
Comment (jja725) [inline job/server/src/main/java/alluxio/master/job/plan/PlanCoordinator.java:280]: would prefer to deal with batch/non-batch situations in DistributedCmdMetrics class and keep here clean.

**Review by jja725 — CHANGES_REQUESTED** [review]
Thanks for the contribution! leave some comments for this round

File: job/common/src/main/java/alluxio/job/plan/migrate/MigrateConfig.java
Comment (luzhang6) [reply job/common/src/main/java/alluxio/job/plan/migrate/MigrateConfig.java:48]: will revert this

File: job/server/src/main/java/alluxio/master/job/metrics/DistributedCmdMetrics.java
Comment (luzhang6) [reply job/server/src/main/java/alluxio/master/job/metrics/DistributedCmdMetrics.java:159]: used that, thanks!

File: job/server/src/main/java/alluxio/master/job/metrics/DistributedCmdMetrics.java
Comment (luzhang6) [reply job/server/src/main/java/alluxio/master/job/metrics/DistributedCmdMetrics.java:228]: done

File: job/server/src/main/java/alluxio/master/job/metrics/DistributedCmdMetrics.java
Comment (luzhang6) [reply job/server/src/main/java/alluxio/master/job/metrics/DistributedCmdMetrics.java:174]: done

File: job/server/src/main/java/alluxio/master/job/plan/PlanCoordinator.java
Comment (luzhang6) [reply job/server/src/main/java/alluxio/master/job/plan/PlanCoordinator.java:280]: done

File: job/server/src/main/java/alluxio/job/plan/migrate/MigrateDefinition.java
Comment (yuzhu) [inline job/server/src/main/java/alluxio/job/plan/migrate/MigrateDefinition.java:71]: why was this changed? for migrate, there is a source and destination right？

File: job/server/src/main/java/alluxio/master/job/metrics/DistributedCmdMetrics.java
Comment (yuzhu) [inline job/server/src/main/java/alluxio/master/job/metrics/DistributedCmdMetrics.java:186]: extra //

**Review by yuzhu — COMMENTED** [review]
Looks good overall, what steps have you done to verify it is working? please add some additional tests.

**Comment by yuzhu** [comment]
@jja725 PTAL as well

File: job/server/src/main/java/alluxio/job/plan/migrate/MigrateDefinition.java
Comment (luzhang6) [reply job/server/src/main/java/alluxio/job/plan/migrate/MigrateDefinition.java:71]: previously changed this name so that in `getFilePathForNonBatchConfig` the file paths can be obtained by the same name of `getFilePath()`, which makes them look consistent. I could revert this change if not necessary.

File: job/server/src/main/java/alluxio/master/job/metrics/DistributedCmdMetrics.java
Comment (luzhang6) [reply job/server/src/main/java/alluxio/master/job/metrics/DistributedCmdMetrics.java:186]: will remove

**Comment by luzhang6** [comment]
> 

Previously I ran the distributed commands manually (except the cancel operation which is short time and hard to do for distributedLoad and distributedCp commands) and check the stats counters, listed the results at the beginning of this PR.

File: job/server/src/main/java/alluxio/job/plan/migrate/MigrateDefinition.java
Comment (yuzhu) [reply job/server/src/main/java/alluxio/job/plan/migrate/MigrateDefinition.java:71]: is there still a `getDestination` function? I imagine that is still useful?

**Comment by yuzhu** [comment]
@luzhang6 please fix checkstyle thanks.

File: job/server/src/main/java/alluxio/job/plan/migrate/MigrateDefinition.java
Comment (luzhang6) [reply job/server/src/main/java/alluxio/job/plan/migrate/MigrateDefinition.java:71]: @yuzhu Yeah, getDestination is untouched. If this change causes confusion still, I will revert it.

Comment (yuzhu) [reply job/server/src/main/java/alluxio/job/plan/migrate/MigrateDefinition.java:71]: that's fine, please add a comment to `getFilePath` if there isn't one already.

Comment (luzhang6) [reply job/server/src/main/java/alluxio/job/plan/migrate/MigrateDefinition.java]: I reverted the change of name this time, will revisit it if there are other new needs to change the name.

**Comment by yuzhu** [comment]
@luzhang6 license header?

File: tests/src/test/java/alluxio/client/hadoop/FileSystemRenameIntegrationTest.java
Comment (jja725) [inline tests/src/test/java/alluxio/client/hadoop/FileSystemRenameIntegrationTest.java:239]: revert this change since it's a known issue when running unit tests locally.

File: tests/src/test/java/alluxio/client/cli/job/DistributedCommandsCancelStatsTest.java
Comment (jja725) [inline tests/src/test/java/alluxio/client/cli/job/DistributedCommandsCancelStatsTest.java:74]: is this cancel operation guaranteed to work before the job is complete?  have the same concern about other unit tests.

**Review by jja725 — CHANGES_REQUESTED** [review]
LGTM overall. Having some questions about the unit test.

File: tests/src/test/java/alluxio/client/cli/job/DistributedCommandsCancelStatsTest.java
Comment (luzhang6) [reply tests/src/test/java/alluxio/client/cli/job/DistributedCommandsCancelStatsTest.java:74]: Yes, i think so, below in the test it will check if the job is cancelled or not, if cancels successfully then the test will pass, otherwise it will fail. Same for other tests.

File: tests/src/test/java/alluxio/client/hadoop/FileSystemRenameIntegrationTest.java
Comment (luzhang6) [reply tests/src/test/java/alluxio/client/hadoop/FileSystemRenameIntegrationTest.java:239]: will revert, thanks.

**Comment by yuzhu** [comment]
@luzhang6 could you check against the test failure sheet and indicate which ones are new and which ones are known?

File: job/server/src/main/java/alluxio/master/job/metrics/DistributedCmdMetrics.java
Comment (yuzhu) [inline job/server/src/main/java/alluxio/master/job/metrics/DistributedCmdMetrics.java:258]: why expand the exception type? what are we including additionally ?

File: tests/src/test/java/alluxio/client/cli/job/DistributedCommandsCancelStatsTest.java
Comment (yuzhu) [reply tests/src/test/java/alluxio/client/cli/job/DistributedCommandsCancelStatsTest.java:74]: I think @jja725 's concern is that the cancel might be a noop if the job has already finished by the time it reaches the cancel call, since the job is very small.

Comment (yuzhu) [reply tests/src/test/java/alluxio/client/cli/job/DistributedCommandsCancelStatsTest.java:74]: in that case, we are not testing anything at all.

File: core/common/src/main/java/alluxio/metrics/MetricKey.java
Comment (yuzhu) [inline core/common/src/main/java/alluxio/metrics/MetricKey.java:741]: one high level comment here, it might be easier if we programmatically construct these MetricKey's so that if we decide to add another JobType it will be easier to add?

**Review by yuzhu — COMMENTED** [review]
Left some comments

File: job/server/src/main/java/alluxio/master/job/metrics/DistributedCmdMetrics.java
Comment (luzhang6) [reply job/server/src/main/java/alluxio/master/job/metrics/DistributedCmdMetrics.java:258]: Expand to catch runtime exceptions including failure to create URI for nulls, etc.

File: tests/src/test/java/alluxio/client/cli/job/DistributedCommandsCancelStatsTest.java
Comment (luzhang6) [reply tests/src/test/java/alluxio/client/cli/job/DistributedCommandsCancelStatsTest.java:74]: If the job has finished, then the output and stat counter values below would be different, and this test would fail. Do I interpret correctly?

File: core/common/src/main/java/alluxio/metrics/MetricKey.java
Comment (luzhang6) [reply core/common/src/main/java/alluxio/metrics/MetricKey.java:741]: Wrap up in this PR or leave it to be done in a next PR when adding for new JobTypes?

Comment (yuzhu) [reply core/common/src/main/java/alluxio/metrics/MetricKey.java:741]: sounds good.

File: tests/src/test/java/alluxio/client/cli/job/DistributedCommandsCancelStatsTest.java
Comment (yuzhu) [reply tests/src/test/java/alluxio/client/cli/job/DistributedCommandsCancelStatsTest.java:74]: I see, then it depends on the timing of the test， which may lead to flakey test.  Please add a comment to the test indicating that some timing assumption is used in this test.

**Review by yuzhu — APPROVED** [review]
LGTM, once the comment for the test is added.

File: tests/src/test/java/alluxio/client/cli/job/DistributedCommandsCancelStatsTest.java
Comment (luzhang6) [reply tests/src/test/java/alluxio/client/cli/job/DistributedCommandsCancelStatsTest.java:74]: Yes, it depends on timing of the test, will add such a comment.

**Review by jja725 — APPROVED** [review]
LGTM, except the concern about the cancel timing. We may want to revisit how we can handle cancel properly if it's flaky.

File: tests/src/test/java/alluxio/client/cli/job/DistributedCommandsCancelStatsTest.java
Comment (luzhang6) [reply tests/src/test/java/alluxio/client/cli/job/DistributedCommandsCancelStatsTest.java:80]: Done.

**Comment by yuzhu** [comment]
alluxio-bot, merge this please

**Pull request merged**
